File: sdk-go/internal/core/url.go
package core

import (
	"net/url"
	"sort"
	"strings"
)

// BuildURL concatenates baseURL + apiPath + path and appends a query string.
// Pure function; no I/O. Trailing slashes on baseURL are trimmed; missing
// leading slashes on path are added.
//
// Query values are stable-sorted by key for deterministic output.
func BuildURL(baseURL, apiPath, path string, query map[string]string) string {
	base := strings.TrimRight(baseURL, "/")
	p := path
	if !strings.HasPrefix(p, "/") {
		p = "/" + p
	}

	out := base + apiPath + p

	if len(query) == 0 {
		return out
	}

	keys := make([]string, 0, len(query))
	for k := range query {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	values := url.Values{}
	for _, k := range keys {
		v := query[k]
		if v == "" {
			continue
		}
		values.Add(k, v)
	}
	if encoded := values.Encode(); encoded != "" {
		out += "?" + encoded
	}
	return out
}

File: sdk-python/src/threecommon/_core/url.py
"""URL builder. Pure function — no I/O."""

from __future__ import annotations

from urllib.parse import urlencode


def build_url(
    *,
    base_url: str,
    api_path: str,
    path: str,
    query: dict[str, str] | None = None,
) -> str:
    """Concatenate ``base_url + api_path + path`` and append a sorted query string.

    Trailing slashes on ``base_url`` are trimmed; missing leading slashes on
    ``path`` are added. Query keys are stable-sorted for deterministic output.
    """
    base = base_url.rstrip("/")
    p = path if path.startswith("/") else "/" + path
    out = base + api_path + p

    if not query:
        return out

    pairs = [(k, v) for k, v in sorted(query.items()) if v]
    if not pairs:
        return out

    return f"{out}?{urlencode(pairs)}"
